// @heavy-duty/spl-utils/src/transfer/transfer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transfer = void 0;
const tslib_1 = require("tslib");
const solana_utils_1 = require("@heavy-duty/solana-utils");
const create_transfer_instructions_1 = require("./create-transfer-instructions");
function transfer(params) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const instructions = (0, create_transfer_instructions_1.createTransferInstructions)({
            amount: params.amount,
            mintAddress: params.mintAddress,
            receiverAddress: params.receiverAddress,
            senderAddress: params.payerKeypair.publicKey.toBase58(),
            fundReceiver: params.fundReceiver,
            memo: params.memo,
        });
        const { transaction, latestBlockhash } = yield (0, solana_utils_1.createTransaction)({
            payerAddress: params.payerKeypair.publicKey.toBase58(),
            connection: params.connection,
            instructions,
            signers: [params.payerKeypair],
        });
        const signature = yield (0, solana_utils_1.sendAndConfirmTransaction)({
            connection: params.connection,
            latestBlockhash,
            transaction,
        });
        return signature;
    });
}
exports.transfer = transfer;
